### ISSUE TITLE
[Bug]: Update Document in VectorStoreIndex create a new Document

### ISSUE THREAD

**Issue by johngonzalez** [run-llama/llama_index#7025]
### Bug Description

A.n unexpected response is obtained. When I use `index.update_ref_doc(doc)` not is updated  a document, really is created a new doc, but if you call it for second time a error is throwed

### Version

llama_index-0.7.11.post1

### Steps to Reproduce

Here code to reproduce:

```
from llama_index import Document, VectorStoreIndex
# Create index store from documents (instances of Document), you can reproduce it with your own documents
index = VectorStoreIndex.from_documents(documents)
# update text for first document
documents[0].text = documents[0].text + ' Any thing'
# When is called for first time create a new document. When is called for second time throw error
index.update_ref_doc(documents[0])
```

### Relevant Logs/Tracbacks

```shell
# Error in second call of update_ref_doc method

KeyError                                  Traceback (most recent call last)
<ipython-input-162-46c0a6b834a7> in <cell line: 2>()
      1 # Borramos
----> 2 index.update_ref_doc(documents[0])

2 frames
/usr/local/lib/python3.10/dist-packages/llama_index/data_structs/data_structs.py in delete(self, doc_id)
    196     def delete(self, doc_id: str) -> None:
    197         """Delete a Node."""
--> 198         del self.nodes_dict[doc_id]
    199 
    200     @classmethod

KeyError: '2b88cc69-5565-4a20-b1d6-902cf7939ef5'
```

**Comment by logan-markewich** [comment]
Hmm, I'm pretty sure this might be related to the fact that changing the text directly does not change the hash of the content. I wonder if I can configure a pydantic validator to run after the text is changed and update the hash 🤔

**Comment by tyre** [comment]
This happens because `VectorStoreIndex.update_ref_doc` deletes the original document and then re-adds its nodes. When it deletes, it will remove the reference from the index store (where the embeddings are held) but not from the docstore (where the documents are held).

@logan-markewich  is that the correct distinction between an index and document store?

`delete_ref_doc` takes a `delete_from_docstore` parameter that will also propagate the deletion to the document store. It defaults to `False` in the `BaseIndex` implementation and, since `VectorStoreIndex` does not define its own implementation of `update_ref_doc`, it will remain `False`. This "fixes" it:

```python
class VectorStoreIndex(BaseIndex[IndexDict]):
    def update_ref_doc(self, document: Document, **update_kwargs: Any) -> None:
        return super().update_ref_doc(
            document, delete_kwargs={"delete_from_docstore": True}, **update_kwargs
        )
``` 

tl;dr, passing that argument down through `update_ref_doc` will cause the docstore to also have the old document removed, but I don't know why the option exists.

@logan-markewich is this a bug or intended behavior? When would one want to delete form the docstore vs. not?

**Comment by logan-markewich** [comment]
@tyre it's intended, because a docstore can be re-used across indexes.

However, the index store/index struct keeps track of which ids are available to an index.

So, if two indexes are sharing the docstore, and a document gets deleted, the other index will encounter errors.

So, that leads us to this default behavior. I agree it's not intuitive, but, I'm not sure how to best handle it 🤔 maybe if the second index encounters a stale node I'd that was deleted, it should just gracefully handle that and remove it?
